internal/terraform/context_apply_deferred_test.go

```diff
@@ -1813,22 +2229,38 @@ func (provider *deferredActionsProvider) Provider() providers.Interface {
 			}
 
 			plannedState := req.ProposedNewState
+			if key == "deferred_resource_change" {
+				deferred = &providers.Deferred{
+					Reason: providers.DeferredReasonProviderConfigUnknown,
+				}
+			}
+
 			if plannedState.GetAttr("output").IsNull() {
 				plannedStateValues := req.ProposedNewState.AsValueMap()
 				plannedStateValues["output"] = cty.UnknownVal(cty.String)
 				plannedState = cty.ObjectVal(plannedStateValues)
+			} else if plannedState.GetAttr("output").AsString() == "mark_for_replacement" {
+				requiresReplace = append(requiresReplace, cty.GetAttrPath("name"), cty.GetAttrPath("output"))
 			}
 
 			provider.plannedChanges.Set(key, plannedState)
 			return providers.PlanResourceChangeResponse{
-				PlannedState: plannedState,
+				PlannedState:    plannedState,
+				Deferred:        deferred,
+				RequiresReplace: requiresReplace,
 			}
 		},
 		ApplyResourceChangeFn: func(req providers.ApplyResourceChangeRequest) providers.ApplyResourceChangeResponse {
-			key := req.Config.GetAttr("name").AsString()
+			var key string
+			if v := req.Config.GetAttr("name"); v.IsKnown() {
+				key = req.Config.GetAttr("name").AsString()
+			} else {
+				key = "<unknown>"
+			}
 
 			newState := req.PlannedState
-			if !newState.GetAttr("output").IsKnown() {
+
+			if req.PlannedState.IsKnown() && !req.PlannedState.IsNull() && !newState.GetAttr("output").IsKnown() {
```

Comment:
And here, it's okay for `output` to be unknown because it's a computed attribute - we're expecting the provider to generate a value for us. But we shouldn't be applying something in which the entire planned state was unknown, that should indicate an error.

Reply:
Yeah makes sense, I think I saw a panic earlier due to a bad mock value in plan, leading to too much caution here